sound/soc/sof/ipc4-pcm.c

```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
 	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;
 	struct ipc4_pipeline_set_state_data *data;
+	struct snd_soc_dapm_widget_list *list;
 	struct snd_sof_widget *pipe_widget;
+	struct snd_soc_dapm_widget *widget;
 	struct sof_ipc4_pipeline *pipeline;
+	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
+	int num_widgets;
 	int ret;
 	int i, j;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
+        list = spcm->stream[substream->stream].list;
+        for_each_dapm_widgets(list, num_widgets, widget) {
```

Comment:
@jsarha this step is incorrect. Why would you loop through the widgets here and down again in line 130 onwards. What you need is to modify the loop here:
```
for (i = pipeline_list->count - 1; i >= 0; i--) {
		pipe_widget = pipeline_list->pipe_widgets[i];
		pipeline = pipe_widget->private;
		if (pipeline->state != state && !pipeline->skip_during_fe_trigger)
			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
	}
```

to do something like 
```
for (i = pipeline_list->count - 1; i >= 0; i--) {
		pipe_widget = pipeline_list->pipe_widgets[i];
		pipeline = pipe_widget->private;
		if (pipeline->use_chain_dma)
                               return sof_ipc4_chaind_dma(sdev, pipe_widget);
	}
```